Transaction value int64 overflow error (postgres bigint doesn't allow unsigned int)

Sharders are crashing with the error :

Defaulted container "helm-sharder-01" out of: helm-sharder-01, helm-promtail-01
panic: failed to encode args[59]: unable to encode 0xffffffffc50fbdbe into binary format for int8 (OID 20): 18446744072720727486 is greater than maximum value for int64


TestCreateAllocation-Create_allocation_for_locking_cost_less_than_minimum_cost_should_not_work is the cause of this error because it is trying to lock negative tokens to create allocation. But sharders should not panic in such scenarios and we should fix it.

https://0chain.slack.com/archives/C03788G8N2D/p1684651726750149

--- a/code/go/0chain.net/chaincore/chain/handler.go
+++ b/code/go/0chain.net/chaincore/chain/handler.go
@@ -1360,8 +1360,17 @@ func PutTransaction(ctx context.Context, entity datastore.Entity) (interface{}, 
 		return nil, fmt.Errorf("put_transaction: invalid request %T", entity)
 	}
 
-	if txn.CreationDate < common.Now()-common.Timestamp(transaction.TXN_TIME_TOLERANCE) {
-		return nil, fmt.Errorf("put_transaction: time out of sync with server time")
+	err := txn.Validate(ctx)
+	if err != nil {
+		logging.Logger.Error("put transaction error", zap.String("txn", txn.Hash), zap.Error(err))
+		return nil, err
+	}
+
+	if txn.Value > config.MaxTokenSupply {
+		logging.Logger.Error("put transaction error - value exceeds max token supply",
+			zap.Uint64("value", uint64(txn.Value)),
+			zap.Uint64("max_token_supply", config.MaxTokenSupply))
+		return nil, fmt.Errorf("transaction value exceeds max token supply")
 	}
 
 	sc := GetServerChain()

--- a/code/go/0chain.net/chaincore/config/config.go
+++ b/code/go/0chain.net/chaincore/config/config.go
@@ -15,6 +15,9 @@ import (
 	"0chain.net/core/viper"
 )
 
+// MaxTokenSupply represents the maximum token supply in SAS
+const MaxTokenSupply = 4e18 // max token supply 400 million ZCN
+
 var (
 	// SmartContractConfig stores the wrapper Viper global instance
 	// for the blockchain smart contracts configuration system.

--- a/code/go/0chain.net/miner/protocol_block.go
+++ b/code/go/0chain.net/miner/protocol_block.go
@@ -8,6 +8,7 @@ import (
 	"math"
 	"sort"
 	"strconv"
+	"sync"
 	"time"
 
 	cstate "0chain.net/chaincore/chain/state"
@@ -492,15 +493,30 @@ func (mc *Chain) ValidateTransactions(ctx context.Context, b *block.Block) error
 		if numWorkers*mc.ValidationBatchSize() < len(b.Txns) {
 			numWorkers++
 		}
-		aggregate := true
-		var aggregateSignatureScheme encryption.AggregateSignatureScheme
-		if aggregate {
-			aggregateSignatureScheme = encryption.GetAggregateSignatureScheme(mc.ClientSignatureScheme(), len(b.Txns), mc.ValidationBatchSize())
+		var aggregate bool
+		aggregateSignatureScheme := encryption.GetAggregateSignatureScheme(mc.ClientSignatureScheme(), len(b.Txns), mc.ValidationBatchSize())
+		if aggregateSignatureScheme != nil {
+			aggregate = true
 		}
-		if aggregateSignatureScheme == nil {
-			aggregate = false
+
+		var (
+			validChannel   = make(chan bool, numWorkers)
+			buildInTxnsMap = make(map[string]struct{})
+			bicLock        = &sync.Mutex{}
+		)
+
+		hasDuplicateBuildInTxns := func(txn *transaction.Transaction) bool {
+			bicLock.Lock()
+			defer bicLock.Unlock()
+			if mc.isBuildInTxn(txn) {
+				if _, ok := buildInTxnsMap[txn.FunctionName]; ok {
+					return true
+				}
+				buildInTxnsMap[txn.FunctionName] = struct{}{}
+			}
+			return false
 		}
-		validChannel := make(chan bool, numWorkers)
+
 		validate := func(ctx context.Context, txns []*transaction.Transaction, start int) {
 			result := false
 			defer func() {
@@ -529,6 +545,16 @@ func (mc *Chain) ValidateTransactions(ctx context.Context, b *block.Block) error
 				if err != nil {
 					cancel = true
 					logging.Logger.Error("validate transactions", zap.Int64("round", b.Round), zap.String("block", b.Hash), zap.String("txn", datastore.ToJSON(txn).String()), zap.Error(err))
+					return
+				}
+
+				if hasDuplicateBuildInTxns(txn) {
+					logging.Logger.Error("validate transactions - duplicated build-in transaction",
+						zap.Int64("round", b.Round),
+						zap.String("block", b.Hash),
+						zap.String("txn", txn.Hash),
+						zap.String("function_name", txn.FunctionName))
+					cancel = true
 					return
 				}
 
@@ -1024,7 +1050,7 @@ func (mc *Chain) generateBlock(ctx context.Context, b *block.Block,
 	cctx, cancel := context.WithTimeout(ctx, mc.ChainConfig.BlockProposalMaxWaitTime())
 	defer cancel()
 
-	buildInTxns, cost, err := mc.buildInTxns(ctx, lfb, b, blockState)
+	buildInTxns, cost, err := mc.buildInTxns(ctx, lfb, b)
 	if err != nil {
 		return fmt.Errorf("get build-in txns failed: %v", err)
 	}
@@ -1258,7 +1284,7 @@ l:
 	return nil
 }
 
-func (mc *Chain) buildInTxns(ctx context.Context, lfb, b *block.Block, state util.MerklePatriciaTrieI) ([]*transaction.Transaction, int, error) {
+func (mc *Chain) buildInTxns(ctx context.Context, lfb, b *block.Block) ([]*transaction.Transaction, int, error) {
 	txns := make([]*transaction.Transaction, 0, 4)
 
 	if mc.ChainConfig.IsFeeEnabled() {

--- a/code/go/0chain.net/miner/protocol_block.go
+++ b/code/go/0chain.net/miner/protocol_block.go
@@ -918,6 +918,14 @@ func txnIterHandlerFunc(
 			return false, nil
 		}
 
+		if txn.Value > config.MaxTokenSupply {
+			logging.Logger.Error("generate block, invalid transaction value",
+				zap.String("hash", txn.Hash),
+				zap.Uint64("value", uint64(txn.Value)))
+			tii.invalidTxns = append(tii.invalidTxns, txn)
+			return false, errors.New("invalid transaction value, exceeds max token supply")
+		}
+
 		cost, fee, err := mc.EstimateTransactionCostFee(ctx, lfb, txn, chain.WithSync(), chain.WithNotifyC(waitC))
 		if err != nil {
 			logging.Logger.Debug("Bad transaction cost fee",

--- a/code/go/0chain.net/miner/transaction.go
+++ b/code/go/0chain.net/miner/transaction.go
@@ -1,0 +1,27 @@
+package miner
+
+import "0chain.net/chaincore/transaction"
+
+const (
+	payFeesTxnName               = "payFees"
+	commitSettingsChangesTxnName = "commit_settings_changes"
+	blobberBlockRewardsTxnName   = "blobber_block_rewards"
+	generateChallengeTxnName     = "generate_challenge"
+)
+
+var gBuildInTxnsMap = map[string]struct{}{
+	payFeesTxnName:               {},
+	commitSettingsChangesTxnName: {},
+	blobberBlockRewardsTxnName:   {},
+	generateChallengeTxnName:     {},
+}
+
+// isBuildInTxn checks if the txn is build-in txn.
+func (mc *Chain) isBuildInTxn(txn *transaction.Transaction) bool {
+	if txn.TransactionType != transaction.TxnTypeSmartContract {
+		return false
+	}
+
+	_, ok := gBuildInTxnsMap[txn.FunctionName]
+	return ok
+}

